### PR TITLE
Add support for the static assets

### PR DESCRIPTION
**Pull request by zubair-arbi**
Added basic settings to support static assets.

@clintonb @cpennington

### CODE REVIEW

File: {{cookiecutter.repo_name}}/.gitignore [hunk 1]
```diff
@@ -30,6 +30,7 @@ default.db
 # Static assets
 media/cache/
 assets/
+*/static/css/
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/.gitignore]: Why is this necessary?

File: {{cookiecutter.repo_name}}/Makefile [hunk 2]
```diff
@@ -1,10 +1,12 @@
 .DEFAULT_GOAL := test
+NODE_BIN=./node_modules/.bin
 
 .PHONY: clean compile_translations dummy_translations extract_translations fake_translations help html_coverage \
 	migrate pull_translations push_translations quality requirements test update_translations validate
 
 help:
-	@echo "Please use \`make <target>' where <target> is one of"
+	@echo "Please use \`make <target>\` where <target> is one of"
+	@echo "  bundle                     build optimized JS bundle, compile CSS, and run collectstatic"
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile]: This target doesn't exist. If it did, it seems it does the same thing as `static`.

File: {{cookiecutter.repo_name}}/Makefile [hunk 3]
```diff
@@ -36,14 +48,30 @@ requirements:
 	pip install -qr requirements.txt --exists-action w
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile:45]: @cpennington I vote we return to using the `requirements` target for local development, and `prod-requirements` for production. You've broken my muscle memory!

File: {{cookiecutter.repo_name}}/Makefile [hunk 4]
```diff
@@ -36,14 +48,30 @@ requirements:
...
 
 test: clean
-	coverage run ./manage.py test {{cookiecutter.repo_name}} --settings={{cookiecutter.repo_name}}.settings.test
+	REUSE_DB=1 coverage run ./manage.py test {{cookiecutter.repo_name}} --settings={{cookiecutter.repo_name}}.settings.test
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile]: Revert. `REUSE_DB` doesn't work anymore.

File: {{cookiecutter.repo_name}}/requirements/base.txt [hunk 5]
```diff
@@ -1,6 +1,9 @@
 django>=1.8.7,<1.9
+django_compressor==1.5
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/requirements/base.txt]: Remove the version.

File: {{cookiecutter.repo_name}}/requirements/base.txt [hunk 6]
```diff
@@ -1,6 +1,9 @@
...
 django-extensions==1.5.5
+django-libsass==0.4
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/requirements/base.txt]: Remove the version.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 7]
```diff
@@ -30,11 +30,14 @@
     'rest_framework',
     'social.apps.django_app.default',
     'waffle',
+    'rest_framework_swagger',
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: Add swagger in a follow-up PR. Keep this one dedicated solely to static assets.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 8]
```diff
@@ -30,11 +30,14 @@
...
+    'compressor',
 )
 
 PROJECT_APPS = (
     '{{cookiecutter.repo_name}}.apps.core',
     '{{cookiecutter.repo_name}}.apps.api',
+    '{{cookiecutter.repo_name}}.apps.{{cookiecutter.repo_name}}',
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: This is not necessary.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 9]
```diff
@@ -89,6 +92,17 @@
     root('conf', 'locale'),
 )
 
+# DRF settings
+REST_FRAMEWORK = {
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: Again, keep this PR dedicated solely to static assets. Please remove this.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 10]
```diff
@@ -161,6 +189,40 @@
 ENABLE_AUTO_AUTH = False
 AUTO_AUTH_USERNAME_PREFIX = 'auto_auth_'
 
+OAUTH2_PROVIDER_URL = None
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: Remove.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 11]
```diff
@@ -170,8 +232,16 @@
 # Request the user's permissions in the ID token
 EXTRA_SCOPE = ['permissions']
 
-# TODO Set this to another (non-staff, ideally) path.
-LOGIN_REDIRECT_URL = '/admin/'
+LOGIN_REDIRECT_URL = '/api/v1/{{cookiecutter.repo_name}}/'
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: Revert.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/production.py [hunk 12]
```diff
@@ -13,6 +13,15 @@
 
 ALLOWED_HOSTS = ['*']
 
+# Enable offline compression of CSS/JS
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/production.py]: Make these the default values in `base.py`. Set them to development-friendly values in `local.py`.

**Comment by clintonb** [comment]
Please pull in the [asset pipeline docs from ecommerce](https://github.com/edx/ecommerce/tree/master/docs).

File: {{cookiecutter.repo_name}}/Makefile (hunk 3 above)
Comment (cpennington) [inline {{cookiecutter.repo_name}}/Makefile:45]: Oops! That's ok by me.

File: {{cookiecutter.repo_name}}/Makefile [hunk 13]
```diff
@@ -27,7 +32,14 @@ help:
...
 	coverage erase
-	rm -rf assets
+	rm -rf {{cookiecutter.repo_name}}/assets {{cookiecutter.repo_name}}/static/build coverage htmlcov
+
+requirements.js:
+	npm install
+	$(NODE_BIN)/bower install
+
+requirements: requirements.js
+	pip install -qr requirements/local.txt --exists-action w
```
Comment (cpennington) [inline {{cookiecutter.repo_name}}/Makefile]: You don't need to add a new `requirements` target, you just need to add a new dependency in the existing targets below.

File: {{cookiecutter.repo_name}}/Makefile [hunk 14]
```diff
@@ -27,7 +32,14 @@ help:
...
 	find . -name '*.pyc' -delete
 	coverage erase
-	rm -rf assets
+	rm -rf {{cookiecutter.repo_name}}/assets {{cookiecutter.repo_name}}/static/build coverage htmlcov
+
+requirements.js:
+	npm install
+	$(NODE_BIN)/bower install
+
+requirements: requirements.js
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile]: Combine this with local-requirements. I'll deal with local vs. prod later since it will involve a configuration change.

File: {{cookiecutter.repo_name}}/.gitignore (hunk 1 above)
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/.gitignore]: Not necessary. Removing.

File: {{cookiecutter.repo_name}}/requirements/base.txt (hunk 6 above)
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/requirements/base.txt]: Is this so that the cookie cutter always get the latest package? if so the should i remove the versions from all apps (except django)?

Comment (clintonb) [inline {{cookiecutter.repo_name}}/requirements/base.txt]: That is correct.

**Comment by zubair-arbi** [comment]
@clintonb @cpennington I have addressed the comments please review.
The exception `AttributeError: 'Module' object has no attribute '_locals'` which is showing on this PR for `pylint --rcfile=pylintrc repo_name *.py` is also failing for `master` branch too on my local machine . Any idea how to fix it?

File: {{cookiecutter.repo_name}}/Makefile [hunk 15]
```diff
@@ -27,12 +32,16 @@ help:
...
-local-requirements:
+requirements.js:
+	npm install
+	$(NODE_BIN)/bower install
+
+requirements: requirements.js
 	pip install -qr requirements/local.txt --exists-action w
 
-requirements:
+prod-requirements:
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile:44]: This needs to be added to the help target above.

File: {{cookiecutter.repo_name}}/Makefile [hunk 16]
```diff
@@ -43,7 +52,26 @@ quality:
 	pep8 --config=.pep8 {{cookiecutter.repo_name}} *.py
 	pylint --rcfile=pylintrc {{cookiecutter.repo_name}} *.py
 
-validate: test quality
+serve:
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile]: Why was this added?

File: {{cookiecutter.repo_name}}/Makefile [hunk 17]
```diff
@@ -43,7 +52,26 @@ quality:
...
+	python manage.py runserver 0.0.0.0:{{cookiecutter.port}}
+
+static:
+	$(NODE_BIN)/r.js -o build.js
+	python manage.py collectstatic --noinput -v0
+	python manage.py compress -v0
+
+bundle:
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/Makefile]: Remove.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/local.py [hunk 18]
```diff
@@ -2,6 +2,10 @@
 
 DEBUG = True
 
+# Enable offline compression of CSS/JS
+COMPRESS_ENABLED = False
+COMPRESS_OFFLINE = False
+
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/local.py:8]: Add `COMPRESS_CSS_FILTERS = ['compressor.filters.css_default.CssAbsoluteFilter']`.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 19]
```diff
@@ -108,9 +109,32 @@
...
+STATICFILES_FINDERS = (
+    'django.contrib.staticfiles.finders.FileSystemFinder',
+    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
+    'compressor.finders.CompressorFinder',
+)
+
+COMPRESS_PRECOMPILERS = (
+    ('text/x-scss', 'django_libsass.SassCompiler'),
+)
+COMPRESS_CSS_FILTERS = ['compressor.filters.css_default.CssAbsoluteFilter']
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: Remove.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py [hunk 20]
```diff
@@ -108,9 +109,32 @@
...
+# END STATIC FILE CONFIGURATION
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/base.py]: Move this after the last `COMPRESS_` setting.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/test.py [hunk 21]
```diff
@@ -30,3 +30,13 @@
     },
 }
 # END IN-MEMORY TEST DATABASE
+
+
+# Enable offline compression of CSS/JS
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/settings/test.py]: None of this is necessary, since these are the defaults set in `base.py`.

File: {{cookiecutter.repo_name}}/Makefile (hunk 16 above)
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/Makefile]: To run the created django app. Since we ask for port while initializing cookie cutter on which the app should run so thats why I added it here.

File: {{cookiecutter.repo_name}}/Makefile (hunk 15 above)
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/Makefile:44]: Done

**Comment by zubair-arbi** [comment]
@clintonb Please review.

**Comment by zubair-arbi** [comment]
@feanil @clintonb  It looks like there is only `PhantomJS` available and tests are failing since there is no `Firefox` or `Chrome` for Travis.

File: {{cookiecutter.repo_name}}/Makefile (hunk 16 above)
Comment (cpennington) [inline {{cookiecutter.repo_name}}/Makefile]: We're moving to docker for devstack (and local development), which is what the port is being used for. See the `start-devstack` target. Do we need a target for a local run-server as well?

File: {{cookiecutter.repo_name}}/.travis.yml [hunk 22]
```diff
@@ -2,18 +2,25 @@ language: python
...
     - $HOME/.cache/pip
+    - node_modules
+    - {{cookiecutter.repo_name}}/static/bower_components
 before_install:
     - "export DISPLAY=:99.0"
     - "sh -e /etc/init.d/xvfb start"
 install:
     - pip install -U pip wheel codecov
     - pip install -r requirements/test.txt
+    - make requirements.js
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/.travis.yml]: This and the line above can be replaced with `make requirements`.

File: {{cookiecutter.repo_name}}/karma.conf.js [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+
+    // enable / disable watching file and executing tests whenever any file changes
+    autoWatch: false,
+
+
+    // start these browsers
+    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
+    browsers: ['Firefox', 'Chrome', 'PhantomJS'],
+
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/karma.conf.js]: Just use firefox.

File: {{cookiecutter.repo_name}}/karma.conf.js [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    logLevel: config.LOG_INFO,
+
+
+    // enable / disable watching file and executing tests whenever any file changes
+    autoWatch: false,
+
+
+    // start these browsers
+    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
+    browsers: ['Firefox', 'Chrome', 'PhantomJS'],
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/karma.conf.js]: Just use Firefox.

File: {{cookiecutter.repo_name}}/karma.conf.js [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+    preprocessors: {
+        '{{cookiecutter.repo_name}}/static/js/!(test)/**/*.js': ['coverage']
+    },
+
+    // enabled plugins
+    plugins:[
+       'karma-jasmine',
+       'karma-requirejs',
+       'karma-firefox-launcher',
+       'karma-chrome-launcher',
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/karma.conf.js]: Remove.

File: {{cookiecutter.repo_name}}/karma.conf.js [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+       'karma-phantomjs-launcher',
```
Comment (clintonb) [inline {{cookiecutter.repo_name}}/karma.conf.js]: Remove.

File: {{cookiecutter.repo_name}}/.travis.yml [hunk 27]
```diff
@@ -2,18 +2,25 @@ language: python
 python:
     - "2.7"
 sudo: false
+addons:
+  firefox: "39.0"
```
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/.travis.yml]: @benpatterson Still no Luck. The build failed again as Firefox failed to start.

File: {{cookiecutter.repo_name}}/karma.conf.js (hunk 24 above)
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/karma.conf.js]: This is just for testing. Once Firefox starts working I will remove others.

**Comment by clintonb** [comment]
@zubair-arbi use PhantomJS for now to get your repo initialized. File a ticket to fix Firefox support ASAP.

**Comment by zubair-arbi** [comment]
@clintonb Ok I am going to add just `PhantomJS` for now.
Here is the ticket for browser issue: https://openedx.atlassian.net/browse/DEVOPS-3382

File: {{cookiecutter.repo_name}}/Makefile (hunk 16 above)
Comment (zubair-arbi) [inline {{cookiecutter.repo_name}}/Makefile]: I am going to remove it since we can add it on relevant project if needed.

**Comment by andy-armstrong** [comment]
@zubair-arbi Thanks so much for doing this. We had been planning to do the same thing from the lessons we learned from building the UI Toolkit:

https://github.com/edx/edx-ui-toolkit

A few questions/thoughts:
1. @talbs is investigating dropping Bower in place of using npm. It seems that Bower is losing support and it appears that npm provides most if not all of the same capabilities: 
https://gofore.com/ohjelmistokehitys/stop-using-bower/
2. I'm not sure that it makes sense to add gulp into the cookiecutter in addition to using Makefiles. I personally like gulp but it seems duplicative to have so many different build tools (Makefile, gulp, bower, npm). It would be good to as a team decide what set we should use before codifying it in the cookie-cutter.
3. If we do decide to use gulp, I like the pattern of having each gulp task live in its own file. You can see how @dsjen set this up in the UI Toolkit:
https://github.com/edx/edx-ui-toolkit/blob/master/gulpfile.js
4. It would be good to standardize on the names of the gulp tasks and/or Makefile targets that we are using across repos. We should have a discussion as to what those targets should be.

FYI @AlasdairSwan @dsjen @dan-f @talbs

**Comment by clintonb** [comment]
@zubair-arbi I removed your commit pinning pylint, as well as the pylint and astroid requirements. Those both come from edx-lint. There are a few JS lint issues that need to be addressed.

File: {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/static/js/test/spec-runner.js [hunk 28]
```diff
@@ -0,0 +1,33 @@
+/**
+ * This is where your tests go.  It should happen automatically when you
+ * add files to the karma configuration.
+ */
+
+var specs = [],
+    config = {};
+
+// the E-Commerce application loads gettext identity library via django, thus
```
Comment (dsjen) [inline {{cookiecutter.repo_name}}/{{cookiecutter.repo_name}}/static/js/test/spec-runner.js]: "E-Commerce" should probably be removed from this comment.

**Comment by clintonb** [comment]
@zubair-arbi build your repo using this branch; but, DO NOT merge this PR. We will wait until a consensus is reached regarding static assets.

@andy-armstrong @AlasdairSwan @dsjen @dan-f @talbs what is your ETA?

**Comment by dan-f** [comment]
@clintonb ETA for what? Am I expected to review? I jus saw I was tagged but assumed it was more of an FYI. If you're requesting a review could I get a little more context on this PR?

**Comment by clintonb** [comment]
ETA for developing a solution to IDA static assets.

**Comment by dan-f** [comment]
Ah, got it. I'll let @andy-armstrong (or someone else with more history on the toolkit) answer that as he has more context, but I'm happy to be involved in the discussions of these patterns! So my 2 cents:

The distinction between bower and npm always was a bit confusing to me. If npm suits us, we might as well standardize by _just_ using npm. If we ever need bower, we can bring it in as well.

I like the idea of using Gulp as well as Makefiles. It looks like gulp adds enough value to the UI Toolkit, but necessarily cannot replace the role of Makefiles in our repos.

**Comment by talbs** [comment]
Jumping on the pile here with my $0.02 as well..
- RE: Bower vs. NPM -  I'm happy to simplify within the UXPL and already have an experimental PR out for review (https://github.com/edx/ux-pattern-library/pull/237).
- RE: Gulp - We use Gulp for a lot of our local dev (Sass compile, asset optimization) within the UXPL. I'm happy to continue to consolidate the tasks we use with the UI Toolkit ones (as well as any that @AlasdairSwan has relied upon with previous CC-based apps). If we don't use Gulp, the UXPL will still have the assumptions that each app will handle its own Sass compilation and asset optimization.

**Comment by dsjen** [comment]
2 more cents:
- +1 to the work in dropping bower.  My only concern would be that packages existed on bower and not on NPM, but it seems to be the case that they typically exist on both systems.
- I don't see why gulp couldn't live harmoniously with Make as it seems well suited for executing particular jobs (js linting, js testing, etc.).  Perhaps we could think of it as specialized tooling to support the JS parts.
- +1 to having the gulp tasks separated into their own files (should out to @AlasdairSwan for the original idea)

**Comment by benpatterson** [comment]
Where did Firefox go? Is that still being used?

**Comment by benpatterson** [comment]
Ok I'm catching up now. So...we couldn't figure out how to test on Firefox do we descoped it and backlogged it? Is that how we define 'done'? Sorry to sound harsh but I find this concerning.

**Comment by clintonb** [comment]
@benpatterson https://openedx.atlassian.net/browse/ECOM-3176

Closing this PR until we resolve the front-end build question. If folks can also offer suggestions on RequireJS vs. Webpacks vs. CommonJS, that would be great, too!

**Comment by benpatterson** [comment]
Ah ok so the ticket is a required part of this work. Great.

**Comment by dan-f** [comment]
@benpatterson @clintonb - Check out [this PR](https://github.com/edx/edx-ui-toolkit/pull/2/files#diff-ea0a2ffccfcfafe13879be34fe2d76d5R17) for examples of testing with multiple browsers.

**Comment by clintonb** [comment]
The issue is not with testing multiple browsers. That's easy. The problem is the fact that Firefox is not starting on the Travis machine.

**Comment by dan-f** [comment]
well good luck then!
